cider/util/CiderBitUtils.h

```diff
@@ -183,6 +183,25 @@ inline size_t countSetBits(const uint8_t* bit_vector, size_t end) {
 
   return ans;
 }
+
+inline bool CheckBitVectorEq(const uint8_t* vec1, const uint8_t* vec2, int end) {
+  // do byte-batched check with memcmp
+  size_t bytes = end / 8;
+  if (memcmp(vec1, vec2, bytes)) {
+    return false;
+  }
+
+  // compare remainders
+  size_t i = bytes * 8;
```

Comment:
Yes, thanks for reminding that there is a manual alignment and initialization in [ArrowArrayBuilder](https://github.com/intel/BDTK/blob/main/cider/tests/utils/ArrowArrayBuilder.h#L104). I didn't notice this earlier.

However, since this function is currently used for comparing null buffers, I expected the parameter `end` to be the number of rows in an arrow array, which may not be aligned to 8.

Further, I don't think this alignment and initialization can always be guaranteed, because if one creates the null buffer with `CiderBatch`'s `resizeBatch()` (which will call [resizeNulls()](https://github.com/intel/BDTK/blob/main/cider/tests/utils/ArrowArrayBuilder.h#L104) to resize null buffer), while the allocated memory is aligned to 8, there may be several remaining bits that are not initialized and can be any value (because `resizeNulls()` only initializes the first `size` bits, which may not necessarily align to 8). In this case if we directly `memcmp` the entire byte it is very likely to fail due to the uninitialized bits.

Reply:
make sense. One more thing, arrow array uses [least-significant bit (LSB) numbering]. This means that within a group of 8 bits, it should read right-to-left:

https://arrow.apache.org/docs/format/Columnar.html#encapsulated-message-format